sdks/java/core/src/test/java/org/apache/beam/sdk/util/BufferedElementCountingOutputStreamTest.java

```diff
@@ -156,6 +159,18 @@ public void testWritingBytesWhenFinishedThrows() throws Exception {
     testValues(toBytes("a")).write("b".getBytes());
   }
 
+  @Test
+  public void testBuffersAreTakenAndReturned() throws Exception {
```

Comment:
Since this is now a class with concurrency, it will be good to have a test that exercises that concurrency in a basic way. It doesn't have to do much, but when combined with a dynamic analyzer may protect against introduced bugs.

Reply:
I added boundary testing for the full and empty cases for the buffer pool to complement this sunny-day scenario.